packages/frontend/app/routes/payments._index.tsx

```diff
@@ -31,57 +50,71 @@ export const loader = async ({ request }: LoaderFunctionArgs) => {
     throw json(null, { status: 400, statusText: 'Invalid result.' })
   }
 
-  const { type, ...pagination } = result.data
+  const { type, walletAddressId, ...pagination } = result.data
 
   const payments = await listPayments({
     ...pagination,
-    ...(type ? { filter: { type: { in: type } } } : {})
+    ...(type || walletAddressId
+      ? {
+          filter: {
+            ...(type ? { type: { in: type } } : {}),
+            ...(walletAddressId
+              ? { walletAddressId: { in: [walletAddressId] } }
+              : {})
+          }
+        }
+      : {})
   })
 
-  let previousPageUrl = '',
-    nextPageUrl = ''
-
-  if (payments.pageInfo.hasPreviousPage) {
-    previousPageUrl = `/payments?before=${payments.pageInfo.startCursor}`
-  }
-
-  if (payments.pageInfo.hasNextPage) {
-    nextPageUrl = `/payments?after=${payments.pageInfo.endCursor}`
-  }
-
   return json({
     payments,
-    previousPageUrl,
-    nextPageUrl,
-    type
+    type,
+    walletAddressId: walletAddressId ?? null
   })
 }
 
 export default function PaymentsPage() {
-  const { payments, previousPageUrl, nextPageUrl, type } =
-    useLoaderData<typeof loader>()
-  const setSearchParams = useSearchParams()[1]
-
+  const { payments, type, walletAddressId } = useLoaderData<typeof loader>()
+  const [searchParams, setSearchParams] = useSearchParams()
+  const [walletId, setWalletId] = useState(walletAddressId || '')
```

Comment:
My understanding `walletId` is necessary as a "local" variable in order to make sure it's valid via the form submit before we actually set it in the searchParams?

Reply:
Yes. It is also means we can keep the input field nicely reactive. And if the wallet address is being removed it means we can skip the validation and simply redirect immediately since all we're doing is resetting the paging and removing the wallet address filter. And it makes it easy to capture it from the URL. Sometimes we get here not by form submission but actually by clicking the view paymnents page button when looking at a wallet address and then the walletAddressId is already in the URL. And also makes it easy to remove this value from the input field when we reset the filters.